File: Visitor/python-source/workers/SanTech.py
from __future__ import annotations

from typing import TYPE_CHECKING

from .Visitor import Visitor

if TYPE_CHECKING:
    from rooms import BathRoom
    from rooms import Home
    from rooms import Kitchen
    from rooms import LivingRoom


class SanTech(Visitor):
    """Сантехник: работает только там, где есть вода."""

    def visit_kitchen(self, kitchen: Kitchen) -> None:
        print("  SanTech installs a sink in the kitchen")

    def visit_living_room(self, living_room: LivingRoom) -> None:
        """В гостиной воды нет."""

    def visit_bath_room(self, bath_room: BathRoom) -> None:
        print("  SanTech installs a bath and a sink")

    def visit_home(self, home: Home) -> None:
        print("SanTech came to the house")

File: JavaSource/src/main/java/ru/akhcheck/patterns/visitor/workers/SanTech.java
package ru.akhcheck.patterns.visitor.workers;

import ru.akhcheck.patterns.visitor.rooms.BathRoom;
import ru.akhcheck.patterns.visitor.rooms.Home;
import ru.akhcheck.patterns.visitor.rooms.Kitchen;
import ru.akhcheck.patterns.visitor.rooms.LivingRoom;

/** Сантехник: работает только там, где есть вода. */
public class SanTech implements Visitor {
    @Override
    public void visit(Kitchen kitchen) {
        System.out.println("  SanTech installs a sink in the kitchen");
    }

    @Override
    public void visit(LivingRoom livingRoom) {
        // В гостиной воды нет.
    }

    @Override
    public void visit(BathRoom bathRoom) {
        System.out.println("  SanTech installs a bath and a sink");
    }

    @Override
    public void visit(Home home) {
        System.out.println("SanTech came to the house");
    }
}
